### PR TITLE
AWS Textract

### PR DESCRIPTION
**Pull request by VirginiaDooley**
Ref https://app.asana.com/0/1204880927741389/1205277317580323/f

We can use Textract to:

- shorten our process for one page SoPNs and include HTML SoPNs in the process. 
- accept more file types of any size by first converting to bytes

This work integrates AWS Textract parsing functionality along side our usual SOPN parsing process and saves the output - which may be a structured table or a string depending on the structure of the original SOPN - as a dataframe. 

This PR includes a:

- management command for developing the AWSTextract class to weave into our current sopn parsing process
- helper classes (with tests) to encompass functionality from above

To date, I've managed to successfully parse:

- a 1 page SoPN and convert to a dataframe. I'm confident in the quality of this dataframe to be parsed by existing systems. 
- a 22 page SoPN that started as html but was saved as a PDF before uploading. It's possible that we need to do more parsing work on this dataframe before saving but I didn't want to spend time on a single case at this point. 

With these changes, there was no change to the sopn test tools baseline numbers. This is expected as our current tooling focuses on CamelotParsedSOPNs rather than AWSTextractParsedSOPNs. The next step will be further testing of these changes and improved parsing of `TextractResult` as well as matching of `AWSTextractParsedSOPN` objects.

Screenshot 2023-11-22 at 6 54 28 PM

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1205939296471782

### CODE REVIEW

File: ynr/apps/sopn_parsing/helpers/extract_pages.py [hunk 1]
```diff
@@ -32,3 +38,153 @@ def extract_pages_for_ballot(ballot):
...
+    # keep an eye on this as it may be useful: https://pypi.org/project/amazon-textract-response-parser/
+    # for now, we're just going to use the raw data
+    import pandas as pd
+
+    blocks = textract_result["Blocks"]
+    tables = self.map_blocks(blocks, "TABLE")
+    cells = self.map_blocks(blocks, "CELL")
+    words = self.map_blocks(blocks, "WORD")
+    selections = self.map_blocks(blocks, "SELECTION_ELEMENT")
+    lines = self.map_blocks(blocks, "LINE")
```
Comment (GeoWill) [inline ynr/apps/sopn_parsing/helpers/extract_pages.py:140]: drive by - all the `self.` shouldn't be here I don't think.

Comment (GeoWill) [reply ynr/apps/sopn_parsing/helpers/extract_pages.py:140]: also eg `self.get_children_ids`

File: ynr/apps/sopn_parsing/helpers/extract_pages.py [hunk 2]
```diff
@@ -32,3 +40,194 @@ def extract_pages_for_ballot(ballot):
...
+    textract_result = textract_client.get_document_analysis(JobId=job_id)
+    while textract_result["JobStatus"] not in ["SUCCEEDED", "FAILED"]:
+        sleep(5)
+        textract_result = textract_client.get_document_analysis(JobId=job_id)
+
+    if textract_result["JobStatus"] == "SUCCEEDED":
+        TextractResult.objects.update_or_create(
+            official_document=ballot.sopn,
+            defaults={"job_id": job_id, "json_response": textract_result},
+        )
```
Comment (GeoWill) [inline ynr/apps/sopn_parsing/helpers/extract_pages.py:102]: I've been faffing with textract for another project this week and discovered a bit of a gotcha here. It's possible that the returned result will be truncated. In this case you need to use the next token to get subsequent parts of the response. I didn't really dig into this as I was just trying to hack something together. But the code below did the job. Essentially it copies the initial response, and creates a seperate list of blocks. Goes through all the `NextTokens` and grabs more blocks to shove in the list. Then subs this complete list of blocks into the copy of the original result and returns it. 

```
    if job_status == "SUCCEEDED":
        analysis = copy.deepcopy(result)
        blocks = []
        while True:
            for block in result["Blocks"]:
                blocks.append(block)
            if ("NextToken" not in result.keys()):
                break
            next_token = result["NextToken"]
            result = textract_client.get_document_analysis(
                JobId=start_response["JobId"],
                NextToken=next_token
            )
        analysis.pop("NextToken", None)
        analysis["Blocks"] = blocks
    else:
        analysis = result

    result_json = json.dumps(result, indent=2)
```

File: ynr/apps/sopn_parsing/helpers/extract_pages.py [hunk 3]
```diff
@@ -32,3 +40,194 @@ def extract_pages_for_ballot(ballot):
...
+
+def get_job_results(job_id, ballot):
+    """This is Step 2 of the SOPN parsing process using AWS Textract.
+    https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/textract/client/get_document_analysis.html#
+    """
+
+    textract_result = textract_client.get_document_analysis(JobId=job_id)
+    while textract_result["JobStatus"] not in ["SUCCEEDED", "FAILED"]:
+        sleep(5)
+        textract_result = textract_client.get_document_analysis(JobId=job_id)
```
Comment (GeoWill) [inline ynr/apps/sopn_parsing/helpers/extract_pages.py:96]: Another gotcha here. I think we'd be in danger of getting throttled on Sopn Day. Textract doesn't seem to like to many `GET`s. The recommended solution is to kick off the job, then let textract notify SNS topic (not sure how to wire that up) which sends to a SQS Queue which whoever is doing the get_document_analysis bit pulls from.

Comment (VirginiaDooley) [reply ynr/apps/sopn_parsing/helpers/extract_pages.py:96]: Yes, @symroe and I spoke about this. We have wired it up in this way for this initial implementing/testing phase, but will look at performance next.

File: ynr/apps/sopn_parsing/helpers/extract_pages.py (hunk 2 above)
Comment (VirginiaDooley) [reply ynr/apps/sopn_parsing/helpers/extract_pages.py:102]: Nevermind! Figured it out 🙏

**Comment by symroe** [comment]
This is looking good! There's a lot going on as we migrate from spike to implemtation, so I think the next thing to do is the refactoring so we can think about it a little better, and start testing. 

First a couple of comments:

* There's an empty `helpers.py` in `official_documents`, maybe this is a stub for refactoring?
* I think the whole PR could do with adding some type annotations to make reading it easier, and to help editors when writing new code. For example: `def start_detection(self, test_sopn):` It's not clear what `test_sopn` should be here. Is it an `OfficialDocument` model? A python file object? A file path? Reading the code it seems like it's a file path, but an annotation (and maybe clearer name?) would help.

    I raise this, because defining the interface for this is going to be super important for Future US.

On to the class refactoring. It's possible you've done most of this already (see above about the empty helper file), but I think we want to do more or less the following:

1. Almost everything in the management command can be moved to it's own class. This class can be something like:

```py
class TextractSOPNHelper:  
    def __init__(self, official_document: OfficialDocument, bucket_name: str = None):  
        self.official_document = official_document  
        self.bucket_name = bucket_name or settings.TEXTRACT_S3_BUCKET_NAME  
  
    @property  
    def s3_key(self) -> str:  
        """  
        Return the S3 key for this file (made up of ballot ID?)        """        ...  
  
    def sync_to_s3(self):  
        # Get the S3 key from self.s3_key  
        # upload self.official_document.uploaded_file        # to S3 if it doesn't already exist        ...  
  
    def start_detection(self, replace=False):  
        # If replace==False and existing AWSTextractParsedSOPN, return early.  
        # textract_client.start_document_analysis(...)        # Get JobID        # get or create a AWSTextractParsedSOPN (based on self.official_document)        # Save the jobID on the new AWSTextractParsedSOPN (add this field to the model)        # Return        ...  
  
    def get_job_results(self, blocking=False):  
        COMPLETED_STATES = ("SUCCEEDED", "FAILED", "PARTIAL_SUCCESS")  
        # AWSTextractParsedSOPN.analysis_status should exist and default to  
        # `IN_PROGRESS`        status = self.official_document.AWSTextractParsedSOPN.analysis_status  
        if status in COMPLETED_STATES:  
            return  
        while status not in COMPLETED_STATES:  
            # result = get_document_analysis()  
            # status = results["status"]            # Store raw response on the model            # Save the model            if not blocking:  
                break
```

Out of scope for this class (not this PR though!) is parsing the raw Textract results. That's stored on the model for parsing later on.

Now the management command can be something like:

```py

class Command(BaseSOPNParsingCommand):  
    def add_arguments(self, parser):  
        super().add_arguments(parser)  
        parser.add_argument(  
            "--blocking",  
            action="store_true",  
            help="Wait for AWS to parse each SOPN",  
        )  
  
    def handle(self, *args, **options):  
        queryset = self.get_queryset()  
  
        if options["start_analysis"]:  
            for official_document in queryset:  
                textract_helper = TextractSOPNHelper(official_document)  
                textract_helper.start_detection()  
        if options["get_results"]:  
            for official_document in queryset:  
                textract_helper = TextractSOPNHelper(official_document)  
                textract_helper.get_job_results(options["blocking"])

```

Hopefully that makes some sense. Basically we can split this up quite a bit.

At this stage I think I'd like to see some tests for `TextractSOPNHelper`. It seems [`moto` hasn't implemented `start_document_analysis` or `get_document_analysis` yet](http://docs.getmoto.org/en/latest/docs/services/textract.html), but we can use it for the S3 uploading parts and a standard `mock` response for the rest. I don't think the management command needs testing, if it's as simple as the above. 

Once that's done, I think we can do more or less the same thing for the parsing phase: Have a helper class that does various things relating to parsing and then we can add that to the same management command once the parsing is done. 

Happy to dive in on a call about any of this, but I thnk the above stubs should give a good idea of what I'm thinking.

File: ynr/apps/sopn_parsing/helpers/extract_pages.py [hunk 4]
```diff
@@ -96,6 +97,20 @@ def get_job_results(job_id, ballot):
...
+            next_token = textract_result["NextToken"]
+
+            textract_result = textract_client.get_document_analysis(
+                JobId=job_id, NextToken=next_token
+            )
+        analysis.pop("NextToken", None)
+        analysis["Blocks"] = blocks
         TextractResult.objects.update_or_create(
             official_document=ballot.sopn,
             defaults={"job_id": job_id, "json_response": textract_result},
```
[the remark this answers is not in the document]
Comment (VirginiaDooley) [reply ynr/apps/sopn_parsing/helpers/extract_pages.py:116]: I've changed this to pass in result, it's running super slow now...

File: ynr/apps/sopn_parsing/management/commands/sopn_parsing_aws_textract.py [hunk 5]
```diff
@@ -75,64 +81,104 @@ def get_job_results(self, job_id):
...
+                        if child_id in words
+                        else selections[child_id]["SelectionStatus"]
+                        for child_id in self.get_children_ids(cell)
+                    ]
+                    i = cell["RowIndex"] - 1
+                    j = cell["ColumnIndex"] - 1
+                    content[i][j] = " ".join(cell_contents)
+
+                dataframe = pd.DataFrame(content[1:], columns=content[0])
+                dataframes.append(dataframe)
```
Comment (VirginiaDooley) [inline ynr/apps/sopn_parsing/management/commands/sopn_parsing_aws_textract.py]: @symroe This was the scenario I mentioned on standup. If there are no table blocks found in the Textract result, but there are lines, we join and save as a string before converting to a df, but the result may prove to challenging to match headers, etc.

File: ynr/apps/sopn_parsing/helpers/extract_pages.py [hunk 6]
```diff
@@ -60,6 +62,9 @@ def start_detection(ballot):
         Key=object_key,
         Body=file_bytes,
     )
+    import pdb
```
Comment (symroe) [inline ynr/apps/sopn_parsing/helpers/extract_pages.py]: Debugger

**Review by symroe — APPROVED** [review]

**Comment by symroe** [comment]
yay

**Pull request merged**